### PR TITLE
(fix) Correctly initialise program state fields

### PR DESCRIPTION
**Pull request by CynthiaKamau**
## Requirements

- [ ] This PR has a title that briefly describes the work done including the ticket number. If there is a ticket, make sure your PR title includes a [conventional commit](https://o3-dev.docs.openmrs.org/#/getting_started/contributing?id=your-pr-title-should-indicate-the-type-of-change-it-is) label. See existing PR titles for inspiration.
- [ ] My work conforms to the [OpenMRS 3.0 Styleguide](https://om.rs/styleguide) and [design documentation](https://zeroheight.com/23a080e38/p/880723-introduction).
- [ ] My work includes tests or is validated by existing tests.

## Summary

Prepopulate program states field in different forms for questions with the same program

## Screenshots

Screenshot 2024-08-12 at 13 08 29

### CODE REVIEW

File: src/components/renderer/field/form-field-renderer.component.tsx [hunk 1]
```diff
@@ -76,6 +76,16 @@ export const FormFieldRenderer = ({ field, valueAdapter, repeatOptions }: FormFi
...
+
+    if (field.type === 'programState' && field.questionOptions.enablePreviousValue) {
+      try {
+        context.processor.getHistoricalValue(field, context).then((value) => {
+          setHistoricalValue({value: value.value, display: value.display});
+        });
+      } catch (error) {
+        console.error(error);
+      }
+    }
```
Comment (samuelmale) [inline src/components/renderer/field/form-field-renderer.component.tsx:88]: Seems like historical values are handled in the enclosure or code block above, this seems like a duplication.

Comment (CynthiaKamau) [reply src/components/renderer/field/form-field-renderer.component.tsx:88]: Are you able to see the value when you test on your end?

Comment (samuelmale) [reply src/components/renderer/field/form-field-renderer.component.tsx:88]: Note that this is currently restricted to working in "enter" mode. That could the reason as to why it's not working for you. I however don't object lifting this constraint.

Comment (CynthiaKamau) [reply src/components/renderer/field/form-field-renderer.component.tsx:88]: okay

**Comment by samuelmale** [comment]
The PR title seems ambiguous compared to the underlying changes. @CynthiaKamau, could you please update the PR title to clearly reflect that you are fixing support for "historical value"?

File: src/components/renderer/field/form-field-renderer.component.tsx [hunk 2]
```diff
@@ -148,25 +148,41 @@ export const FormFieldRenderer = ({ field, valueAdapter, repeatOptions }: FormFi
...
-              value={value}
-              errors={errors}
-              warnings={warnings}
-              setFieldValue={(val) => {
-                onChange(val);
-                onAfterChange(val);
-                onBlur();
-              }}
-            />
+            {field.type === 'programState' && historicalValue?.value ? (
```
Comment (samuelmale) [inline src/components/renderer/field/form-field-renderer.component.tsx:159]: I'm afraid of the coupling this potentially introduces. I also don't understand why we are inferring the current value from the "historicalValue". Do you know why `FormFieldValueAdapter.getInitialValue(...)` doesn't suffice? What is the problem we are dealing with here?

Comment (CynthiaKamau) [reply src/components/renderer/field/form-field-renderer.component.tsx:159]: The value of `FormFieldValueAdapter.getInitialValue(...)` is undefined, thats why i used this approach. The intention is to pre-populate the field as opposed to having the user click the `usePreviousValue` button. @slubwama can provide more context on the issue

Comment (samuelmale) [reply src/components/renderer/field/form-field-renderer.component.tsx:159]: > The value of FormFieldValueAdapter.getInitialValue(...) is undefined, thats why i used this approach.

Shouldn't be investigating why the value is undefined? My concern is ensuring that we consistently manage value binding.

**Comment by gracepotma** [comment]
This is a high priority for the METS team for UgandaEMR+. (Thanks for mentioning this on the O3 Squad call today @slubwama; always great to know that work will be used right away!)

@CynthiaKamau were you able to address Male's comments - i.e. are you ready for a re-review?

**Comment by CynthiaKamau** [comment]
> This is a high priority for the METS team for UgandaEMR+. (Thanks for mentioning this on the O3 Squad call today @slubwama; always great to know that work will be used right away!)
> 
> @CynthiaKamau were you able to address Male's comments - i.e. are you ready for a re-review?

on it

**Pull request merged**